markov-analysis.py

```diff
@@ -0,0 +1,220 @@
+'''
+ Exercise 13.8. Markov analysis:
+
+ 1. Write a program to read a text from a file and perform Markov analysis. The 
+    result should be a dictionary that maps from prefixes to a collection of 
+    possible suffixes. The collection might be a list, tuple, or dictionary; it 
+    is up to you to make an appropriate choice. You can test your program with 
+    prefix length two, but you should write the program in a way that makes it 
+    easy to try other lengths.
+
+ 2. Add a function to the previous program to generate random text based on the 
+   Markov analysis. Here is an example from Emma with prefix length 2:
+    He was very clever, be it sweetness or be angry, ashamed or only amused, at 
+    such a stroke. She had never thought of Hannah till you were never meant for
+    me?" "I cannot make speeches, Emma:" he soon cut it all himself.
+
+    For this example, I left the punctuation attached to the words. The result 
+    is almost syntactically correct, but not quite. Semantically, it almost 
+    makes sense, but not quite.
+    What happens if you increase the prefix length? Does the random text make 
+    more sense?
+
+ 3. Once your program is working, you might want to try a mash-up: if you 
+    combine text from two or more books, the random text you generate will blend
+    the vocabulary and phrases from the sources in interesting ways.
+
+ Credit: This case study is based on an example from Kernighan and Pike, The 
+ Practice of Programming, Addison-Wesley, 1999.
+
+ You should attempt this exercise before you go on; then you can can download my
+ solution from http://thinkpython2.com/code/markov.py. You will also need http:
+ //thinkpython2.com/code/emma.txt.
+'''
+
+import random
+import string
+
+'''
+    idea:
+    mapping of n-tuple (of n prefixes) to some collection that has suffixes with 
+    their respective frequencies
+
+    e.g. 
+
+    (p_1, p_2, ..., p_n) -> ((s_1, f_1), (s_2, f_2), ..., (s_n, f_n))
+    where the order of p matters
+    ideally the f_i's would be normalized but iirc doing that to randomly pick a 
+    float x between 0 and 1 might result in something not random, so i think i 
+    will go with the cumulative sum solution from the preview problem
+    another solution might be to just construct a list of each occurrence 
+
+    simple idea:
+    don't use word frequencies, just a set of suffixes
+    so the dict is a mapping P -> S 
+    (p_1, p_2, ..., p_n) -> (s_1, s_2, ..., s_n)
+
+    after finishing:
+    not sure why i ever thought of using a set of suffixes instead of a list,
+    since that would even preserve relative frequencies!
```

Comment:
right.

But what would happen if you used the list approach on a very large corpus, compared to just storing the frequencies directly?

Reply:
that would take up a lot of space. it would be worth implementing `prefix_map` as a map of n-tuples to some collection of suffixes with corresponding frequencies; then, `random_text` would use these frequencies to make a weighted random choice of suffix.